lib/attrsets.nix

```diff
@@ -651,6 +651,20 @@ rec {
     f:
     listToAttrs (map (n: nameValuePair n (f n)) names);
 
+  /* Generate an attribute set by mapping over a list of items
+     Taking two functions, one to apply for the name and one to
+     apply for the value of the attribute set.
+
+    Example:
+       genAttrs' [ {a = 2; b.c = "a" } {a = 1; b.c = "b"} ]
+          (i: i.b.c) (i: i.a)
+       => { a = 2; b = 1; }
+
+    Type:
+      genAttrs' :: [ a ] -> (a -> String) -> (a -> Any) -> AttrSet
```

Comment:
A denotationally equivalent function would be

```diff
-genAttrs' :: [ a ] -> (a -> String) -> (a -> Any) -> AttrSet
+genAttrs' :: [ a ] -> (a -> { name :: String; value :: a}) -> AttrSet
```

However, operationally, the latter can use a let binding to share computation between the two attributes.
The prior can not share a variable that depends on the `a`.

Reply:
Yet another alternative is

```
genAttrs' :: [ a ] -> (a -> Attrs) -> Attrs
```

This more compact, more powerful, and you can still read that a name and value are computed.

```nix
genAttrs' [ {a = 2; b.c = "a" } {a = 1; b.c = "b"} ]
  (i: { ${i.b.c} = i.a; })
```

while also allowing filtering behavior:

```nix
genAttrs' [ {a = 2; b.c = "a" } {a = 1; b.c = "b"} ]
  (i: optionalAttrs (i.a > 1) { ${i.b.c} = i.a; })
```

or multiple keys per item

```nix
genAttrs' [ {a = 2; b.c = "a" } {a = 1; b.c = "b"} ]
  (i: {
    ${i.b.c} = i.a;
    "${i.b.c}-squared" = i.a * i.a;
  })
```

Reply:
Whatever `genAttrs'` ends up being, its documentation should refer to the other functions that are similar, and the other functions should refer back.
Some functions I'd think of
- `genAttrs`
- `concatMapAttrs`
- `listToAttrs`

There's probably a more related function I'm not thinking of.